File: Sources/OpenDDSharp.Standard/DDS/TopicBuiltinTopicData.cs
/*********************************************************************
This file is part of OpenDDSharp.

OpenDDSharp is a .NET wrapper for OpenDDS
Copyright (C) 2018 Jose Morato

OpenDDSharp is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

OpenDDSharp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with OpenDDSharp. If not, see <http://www.gnu.org/licenses/>.
**********************************************************************/
using System;
using System.Collections.Generic;
using System.Runtime.InteropServices;

namespace OpenDDSharp.DDS
{
    /// <summary>
    /// Class that contains information about available <see cref="Topic" />s within the system.
    /// </summary>
    /// <remarks>
    /// The DCPSTopic topic communicates the existence of topics by means of the TopicBuiltinTopicData datatype. Each TopicBuiltinTopicData sample in a Domain represents a <see cref="Topic" /> 
    /// in that Domain: a new TopicBuiltinTopicData instance is created when a newly-added <see cref="Topic" />  is enabled. However, the instance is not disposed when a <see cref="Topic" />  is 
    /// deleted by its participant because a topic lifecycle is tied to the lifecycle of a Domain, not to the lifecycle of an individual participant.
    /// An updated TopicBuiltinTopicData sample is written each time a <see cref="Topic" />  modifies one or more of its QoS policy values.
    /// </remarks>
    public struct TopicBuiltinTopicData : IEquatable<TopicBuiltinTopicData>
    {
        #region Fields
        private List<IntPtr> toRelease;
        #endregion

        #region Properties
        /// <summary>
        /// Gets the global unique identifier of the <see cref="Topic" />.
        /// </summary>
        public BuiltinTopicKey Key { get; internal set; }

        /// <summary>
        /// Gets the name of the <see cref="Topic" />.
        /// </summary>
        public string Name { get; internal set; }

        /// <summary>
        /// Gets the type name of the <see cref="Topic" />.
        /// </summary>
        public string TypeName { get; internal set; }

        /// <summary>
        /// Gets the <see cref="DurabilityQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public DurabilityQosPolicy Durability { get; internal set; }

        /// <summary>
        /// Gets the <see cref="DurabilityServiceQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public DurabilityServiceQosPolicy DurabilityService { get; internal set; }

        /// <summary>
        /// Gets the <see cref="DeadlineQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public DeadlineQosPolicy Deadline { get; internal set; }

        /// <summary>
        /// Gets the <see cref="LatencyBudgetQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public LatencyBudgetQosPolicy LatencyBudget { get; internal set; }

        /// <summary>
        /// Gets the <see cref="LivelinessQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public LivelinessQosPolicy Liveliness { get; internal set; }

        /// <summary>
        /// Gets the <see cref="ReliabilityQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public ReliabilityQosPolicy Reliability { get; internal set; }

        /// <summary>
        /// Gets the <see cref="TransportPriorityQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public TransportPriorityQosPolicy TransportPriority { get; internal set; }

        /// <summary>
        /// Gets the <see cref="LifespanQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public LifespanQosPolicy Lifespan { get; internal set; }

        /// <summary>
        /// Gets the <see cref="DestinationOrderQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public DestinationOrderQosPolicy DestinationOrder { get; internal set; }

        /// <summary>
        /// Gets the <see cref="HistoryQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public HistoryQosPolicy History { get; internal set; }

        /// <summary>
        /// Gets the <see cref="ResourceLimitsQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public ResourceLimitsQosPolicy ResourceLimits { get; internal set; }

        /// <summary>
        /// Gets the <see cref="OwnershipQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public OwnershipQosPolicy Ownership { get; internal set; }

        /// <summary>
        /// Gets the <see cref="TopicDataQosPolicy" /> attached to the <see cref="Topic" />.
        /// </summary>
        public TopicDataQosPolicy TopicData { get; internal set; }
        #endregion

        #region Methods
        internal TopicBuiltinTopicDataWrapper ToNative()
        {
            if (toRelease == null)
            {
                toRelease = new List<IntPtr>();
            }

            var data = new TopicBuiltinTopicDataWrapper
            {
                Deadline = Deadline,
                DestinationOrder = DestinationOrder,
                Durability = Durability,
                DurabilityService = DurabilityService,
                Key = Key,
                LatencyBudget = LatencyBudget,
                Lifespan = Lifespan,
                Liveliness = Liveliness,
                Ownership = Ownership,
                Reliability = Reliability,
                History = History,
                ResourceLimits = ResourceLimits,
                TransportPriority = TransportPriority,
            };

            if (Name != null)
            {
                data.Name = Marshal.StringToHGlobalAnsi(Name);
                toRelease.Add(data.Name);
            }

            if (TypeName != null)
            {
                data.TypeName = Marshal.StringToHGlobalAnsi(TypeName);
                toRelease.Add(data.TypeName);
            }

            if (TopicData != null)
            {
                data.TopicData = TopicData.ToNative();
            }

            return data;
        }

        internal void FromNative(TopicBuiltinTopicDataWrapper wrapper)
        {
            Deadline = wrapper.Deadline;
            DestinationOrder = wrapper.DestinationOrder;
            Durability = wrapper.Durability;
            DurabilityService = wrapper.DurabilityService;
            Key = wrapper.Key;
            LatencyBudget = wrapper.LatencyBudget;
            Lifespan = wrapper.Lifespan;
            Liveliness = wrapper.Liveliness;
            Ownership = wrapper.Ownership;
            Reliability = wrapper.Reliability;
            History = wrapper.History;
            ResourceLimits = wrapper.ResourceLimits;
            TransportPriority = wrapper.TransportPriority;

            if (wrapper.Name != IntPtr.Zero)
            {
                Name = Marshal.PtrToStringAnsi(wrapper.Name);
            }
            else
            {
                Name = null;
            }

            if (wrapper.TypeName != IntPtr.Zero)
            {
                TypeName = Marshal.PtrToStringAnsi(wrapper.TypeName);
            }
            else
            {
                TypeName = null;
            }

            if (TopicData == null)
            {
                TopicData = new TopicDataQosPolicy();
            }
            TopicData.FromNative(wrapper.TopicData);
        }

        internal void Release()
        {
            TopicData?.Release();

            foreach (IntPtr ptr in toRelease)
            {
                Marshal.FreeHGlobal(ptr);
            }

            toRelease.Clear();
        }
        #endregion

        #region IEquatable<TopicBuiltinTopicData> Members
        /// <summary>
        /// Indicates whether the current object is equal to another object of the same type.
        /// </summary>
        /// <param name="other">An object to compare with this object.</param>
        /// <returns><see langword="true" /> if the current object is equal to the other parameter; otherwise, <see langword="false" />.</returns>
        public bool Equals(TopicBuiltinTopicData other)
        {
            return Key == other.Key &&
                   Name == other.Name &&
                   TypeName == other.TypeName &&
                   Durability == other.Durability &&
                   DurabilityService == other.DurabilityService &&
                   Deadline == other.Deadline &&
                   LatencyBudget == other.LatencyBudget &&
                   Liveliness == other.Liveliness &&
                   Reliability == other.Reliability &&
                   TransportPriority == other.TransportPriority &&
                   Lifespan == other.Lifespan &&
                   DestinationOrder == other.DestinationOrder &&
                   History == other.History &&
                   ResourceLimits == other.ResourceLimits &&
                   Ownership == other.Ownership &&
                   TopicData == other.TopicData;
        }

        /// <summary>
        /// Determines whether the specified object is equal to the current object.
        /// </summary>
        /// <param name="obj">The object to compare with the current object.</param>
        /// <returns><see langword="true" /> if the specified object is equal to the current object; otherwise, <see langword="false" />.</returns>
        public override bool Equals(object obj)
        {
            return (obj is TopicBuiltinTopicData other) && Equals(other);
        }

        /// <summary>
        /// Serves as the default hash function.
        /// </summary>
        /// <returns>A hash code for the current object.</returns>
        public override int GetHashCode()
        {
            var hashCode = 1476352029;
            hashCode = (hashCode * -1521134295) + Key.GetHashCode();
            hashCode = (hashCode * -1521134295) + EqualityComparer<string>.Default.GetHashCode(Name);
            hashCode = (hashCode * -1521134295) + EqualityComparer<string>.Default.GetHashCode(TypeName);
            hashCode = (hashCode * -1521134295) + EqualityComparer<DurabilityQosPolicy>.Default.GetHashCode(Durability);
            hashCode = (hashCode * -1521134295) + EqualityComparer<DurabilityServiceQosPolicy>.Default.GetHashCode(DurabilityService);
            hashCode = (hashCode * -1521134295) + EqualityComparer<DeadlineQosPolicy>.Default.GetHashCode(Deadline);
            hashCode = (hashCode * -1521134295) + EqualityComparer<LatencyBudgetQosPolicy>.Default.GetHashCode(LatencyBudget);
            hashCode = (hashCode * -1521134295) + EqualityComparer<LivelinessQosPolicy>.Default.GetHashCode(Liveliness);
            hashCode = (hashCode * -1521134295) + EqualityComparer<ReliabilityQosPolicy>.Default.GetHashCode(Reliability);
            hashCode = (hashCode * -1521134295) + EqualityComparer<TransportPriorityQosPolicy>.Default.GetHashCode(TransportPriority);
            hashCode = (hashCode * -1521134295) + EqualityComparer<LifespanQosPolicy>.Default.GetHashCode(Lifespan);
            hashCode = (hashCode * -1521134295) + EqualityComparer<DestinationOrderQosPolicy>.Default.GetHashCode(DestinationOrder);
            hashCode = (hashCode * -1521134295) + EqualityComparer<HistoryQosPolicy>.Default.GetHashCode(History);
            hashCode = (hashCode * -1521134295) + EqualityComparer<ResourceLimitsQosPolicy>.Default.GetHashCode(ResourceLimits);
            hashCode = (hashCode * -1521134295) + EqualityComparer<OwnershipQosPolicy>.Default.GetHashCode(Ownership);
            hashCode = (hashCode * -1521134295) + EqualityComparer<TopicDataQosPolicy>.Default.GetHashCode(TopicData);

            return hashCode;
        }
        #endregion

        #region Operators
        /// <summary>
        /// Equals comparison operator.
        /// </summary>
        /// <param name="left">The left value for the comparison.</param>
        /// <param name="right">The right value for the comparison.</param>
        /// <returns><see langword="true" /> if the left object is equal to the right object; otherwise, <see langword="false" />.</returns>
        public static bool operator ==(TopicBuiltinTopicData left, TopicBuiltinTopicData right)
        {
            return left.Equals(right);
        }

        /// <summary>
        /// Not equals comparison operator.
        /// </summary>
        /// <param name="left">The left value for the comparison.</param>
        /// <param name="right">The right value for the comparison.</param>
        /// <returns><see langword="false" /> if the left object is equal to the right object; otherwise, <see langword="true" />.</returns>
        public static bool operator !=(TopicBuiltinTopicData left, TopicBuiltinTopicData right)
        {
            return !left.Equals(right);
        }
        #endregion
    }

    [StructLayout(LayoutKind.Sequential, CharSet=CharSet.Ansi)]
    internal struct TopicBuiltinTopicDataWrapper
    {
        #region Fields
        public BuiltinTopicKey Key;
        public IntPtr Name;
        public IntPtr TypeName;
        public DurabilityQosPolicyWrapper Durability;
        public DurabilityServiceQosPolicyWrapper DurabilityService;
        public DeadlineQosPolicyWrapper Deadline;
        public LatencyBudgetQosPolicyWrapper LatencyBudget;
        public LivelinessQosPolicyWrapper Liveliness;
        public ReliabilityQosPolicyWrapper Reliability;
        public TransportPriorityQosPolicyWrapper TransportPriority;
        public LifespanQosPolicyWrapper Lifespan;
        public DestinationOrderQosPolicyWrapper DestinationOrder;
        public HistoryQosPolicyWrapper History;
        public ResourceLimitsQosPolicyWrapper ResourceLimits;
        public OwnershipQosPolicyWrapper Ownership;
        public TopicDataQosPolicyWrapper TopicData;
        #endregion
    }
}

File: Sources/OpenDDSharp/TopicBuiltinTopicData.cpp
/*********************************************************************
This file is part of OpenDDSharp.

OpenDDSharp is a .NET wrapper for OpenDDS
Copyright (C) 2018 - 2021 Jose Morato

OpenDDSharp is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

OpenDDSharp is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with OpenDDSharp. If not, see <http://www.gnu.org/licenses/>.
**********************************************************************/
#include "TopicBuiltinTopicData.h"

OpenDDSharp::DDS::BuiltinTopicKey OpenDDSharp::DDS::TopicBuiltinTopicData::Key::get() {
	return key;
};

System::String^ OpenDDSharp::DDS::TopicBuiltinTopicData::Name::get() {
	return name;
};

System::String^ OpenDDSharp::DDS::TopicBuiltinTopicData::TypeName::get() {
	return type_name;
};

OpenDDSharp::DDS::DurabilityQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::Durability::get() {
	return durability;
};

OpenDDSharp::DDS::DurabilityServiceQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::DurabilityService::get() {
	return durability_service;
};

OpenDDSharp::DDS::DeadlineQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::Deadline::get() {
	return deadline;
};

OpenDDSharp::DDS::LatencyBudgetQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::LatencyBudget::get() {
	return latency_budget;
};

OpenDDSharp::DDS::LivelinessQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::Liveliness::get() {
	return liveliness;
};

OpenDDSharp::DDS::ReliabilityQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::Reliability::get() {
	return reliability;
};

OpenDDSharp::DDS::TransportPriorityQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::TransportPriority::get() {
	return transport_priority;
};

OpenDDSharp::DDS::LifespanQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::Lifespan::get() {
	return lifespan;
};

OpenDDSharp::DDS::DestinationOrderQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::DestinationOrder::get() {
	return destination_order;
};

OpenDDSharp::DDS::HistoryQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::History::get() {
	return history;
};

OpenDDSharp::DDS::ResourceLimitsQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::ResourceLimits::get() {
	return resource_limits;
};

OpenDDSharp::DDS::OwnershipQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::Ownership::get() {
	return ownership;
};

OpenDDSharp::DDS::TopicDataQosPolicy^ OpenDDSharp::DDS::TopicBuiltinTopicData::TopicData::get() {
	return topic_data;
};

void OpenDDSharp::DDS::TopicBuiltinTopicData::FromNative(::DDS::TopicBuiltinTopicData native) {
	msclr::interop::marshal_context context;

	key.FromNative(native.key);	

	const char * name_ptr = native.name;
	name = context.marshal_as<System::String^>(name_ptr);

	const char * type_name_ptr = native.type_name;
	type_name = context.marshal_as<System::String^>(type_name_ptr);

	durability = gcnew OpenDDSharp::DDS::DurabilityQosPolicy();
	durability->FromNative(native.durability);

	durability_service = gcnew OpenDDSharp::DDS::DurabilityServiceQosPolicy();
	durability_service->FromNative(native.durability_service);

	deadline = gcnew OpenDDSharp::DDS::DeadlineQosPolicy();
	deadline->FromNative(native.deadline);

	latency_budget = gcnew OpenDDSharp::DDS::LatencyBudgetQosPolicy();
	latency_budget->FromNative(native.latency_budget);

	liveliness = gcnew OpenDDSharp::DDS::LivelinessQosPolicy();
	liveliness->FromNative(native.liveliness);

	reliability = gcnew OpenDDSharp::DDS::ReliabilityQosPolicy();
	reliability->FromNative(native.reliability);

	transport_priority = gcnew OpenDDSharp::DDS::TransportPriorityQosPolicy();
	transport_priority->FromNative(native.transport_priority);

	lifespan = gcnew OpenDDSharp::DDS::LifespanQosPolicy();
	lifespan->FromNative(native.lifespan);

	destination_order = gcnew OpenDDSharp::DDS::DestinationOrderQosPolicy();
	destination_order->FromNative(native.destination_order);

	history = gcnew OpenDDSharp::DDS::HistoryQosPolicy();
	history->FromNative(native.history);

	resource_limits = gcnew OpenDDSharp::DDS::ResourceLimitsQosPolicy();
	resource_limits->FromNative(native.resource_limits);

	ownership = gcnew OpenDDSharp::DDS::OwnershipQosPolicy();
	ownership->FromNative(native.ownership);

	topic_data = gcnew OpenDDSharp::DDS::TopicDataQosPolicy();
	topic_data->FromNative(native.topic_data);
}


::DDS::TopicBuiltinTopicData OpenDDSharp::DDS::TopicBuiltinTopicData::ToNative() {
	::DDS::TopicBuiltinTopicData native;
	msclr::interop::marshal_context context;

	native.key = key.ToNative();		
	native.name = context.marshal_as<const char*>(name);
	native.type_name = context.marshal_as<const char*>(type_name);	
	native.durability = durability->ToNative();
	native.durability_service = durability_service->ToNative();	
	native.deadline = deadline->ToNative();	
	native.latency_budget = latency_budget->ToNative();	
	native.liveliness = liveliness->ToNative();	
	native.reliability = reliability->ToNative();	
	native.transport_priority = transport_priority->ToNative();	
	native.lifespan = lifespan->ToNative();	
	native.destination_order = destination_order->ToNative();	
	native.history = history->ToNative();	
	native.resource_limits = resource_limits->ToNative();	
	native.ownership = ownership->ToNative();	
	native.topic_data = topic_data->ToNative();

	return native;
}